bika/lims/browser/js/coffee/bika.lims.worksheet.coffee

```diff
@@ -0,0 +1,1070 @@
+      $(m_selector).hide()
+    else if method_constraints[0] == 1
+      $(m_selector).show()
+    else if method_constraints[0] == 2
+      # XXX length check of an object??
+      if analysis_constraints.length > 1
+        $(m_selector).hide()
+        method_name = $(m_selector).find("option[value='#{method_uid}']").innerHtml()
+        $(m_selector).after "<span class='method-label' uid='#{analysis_uid}' href='#'>#{method_name}</span>"
+    else if method_constraints[0] == 3
+      $(m_selector).show()
+
+    # We are going to reload the instrument list.
+    # Enable all disabled options from other Instrument lists which has the same
+    # value as old value of this instrument selectbox.
+    ins_old_val = $(i_selector).val()
+    if ins_old_val and ins_old_val != ''
+      $("table.bika-listing-table select.listing_select_entry[field='Instrument'][value!='#{ins_old_val}'] option[value='#{ins_old_val}']").prop "disabled", no
+
+    # Populate instruments list
+    $(i_selector).find("option").remove()
+    if method_constraints[7]
+      $.each method_constraints[7], (key, value) ->
+        if me.is_instrument_allowed(key)
+          $(i_selector).append "<option value='#{key}'>#{value}</option>"
+        else
+          $(i_selector).append "<option value='#{key}' disabled='disabled'>#{value}</option>"
+
+    # None option in instrument selector?
+    if method_constraints[3] == 1
+      $(i_selector).prepend "<option selected='selected' value=''>#{_('None')}</option>"
+
+    # Select the default instrument
+    if me.is_instrument_allowed method_constraints[4]
+      $(i_selector).val method_constraints[4]
+      # Disable this Instrument in the other Instrument SelectBoxes
+      $("table.bika-listing-table select.listing_select_entry[field='Instrument'][value!='#{method_constraints[4]}'] option[value='#{method_constraints[4]}']").prop "disabled", yes
+
+    # Instrument selector visible?
+    if method_constraints[2] == 0
+      $(i_selector).hide()
+    else if method_constraints[2] == 1
+      $(i_selector).show()
+
+    # Allow to edit results?
+    if method_constraints[5] == 0
+      $(".interim input[uid='#{analysis_uid}']").val ""
+      $("input[field='Result'][uid='#{analysis_uid}']").val ""
+      $(".interim input[uid='#{analysis_uid}']").prop "disabled", yes
+      $("input[field='Result'][uid='#{analysis_uid}']").prop "disabled", yes
+    else if method_constraints[5] == 1
+      $(".interim input[uid='#{analysis_uid}']").prop "disabled", no
+      $("input[field='Result'][uid='#{analysis_uid}']").prop "disabled", no
+
+    # Info/Warn message?
+    $(".alert-instruments-invalid[uid='#{analysis_uid}']").remove()
+    if method_constraints[6] and method_constraints[6] != ""
+      $(i_selector).after "<img uid='#{analysis_uid}' class='alert-instruments-invalid' src='#{@get_portal_url()}/++resource++bika.lims.images/warning.png' title='#{method_constraints[6]}'>"
+    $(".amconstr[uid='#{analysis_uid}']").remove()
+
```

Comment:
Indeed, this is the weirdest function of all them. Just a side note, all these "constraints" can be more understandable (or maybe is more confusing, though) if you look to this: https://github.com/senaite/senaite.core/blob/master/bika/lims/tests/halt/test_method_instrument_constraints.py#L116-L184
I know, is hell, but together with https://cloud.githubusercontent.com/assets/832627/15012945/c4bca092-11fc-11e6-90a3-4628e4fa7d08.png it was a good exercise to squeeze the brain a bit and maybe train it for complex sudoku.